src/mbgl/layout/symbol_layout.cpp

```diff
@@ -418,43 +420,57 @@ void SymbolLayout::createBucket(const ImagePositions&, std::unique_ptr<FeatureIn
 
         // Insert final placement into collision tree and add glyphs/icons to buffers
 
-        if (hasText) {
-            const Range<float> sizeData = bucket->textSizeBinder->getVertexSizeData(feature);
-            bucket->text.placedSymbols.emplace_back(symbolInstance.anchor.point, symbolInstance.anchor.segment, sizeData.min, sizeData.max,
-                    symbolInstance.textOffset, symbolInstance.writingModes, symbolInstance.line, CalculateTileDistances(symbolInstance.line, symbolInstance.anchor));
-            symbolInstance.placedTextIndex = bucket->text.placedSymbols.size() - 1;
-            PlacedSymbol& horizontalSymbol = bucket->text.placedSymbols.back();
-
-            bool firstHorizontal = true;
-            for (const auto& symbol : symbolInstance.horizontalGlyphQuads) {
-                size_t index = addSymbol(
-                    bucket->text, sizeData, symbol,
-                    symbolInstance.anchor, horizontalSymbol);
-                if (firstHorizontal) {
-                    horizontalSymbol.vertexStartIndex = index;
-                    firstHorizontal = false;
-                }
+        if (hasText && feature.formattedText) {
+            const auto& formattedText = *feature.formattedText;
+            std::function<void (std::size_t sectionIndex, bool updateLastSection)> updatePaintProperties;
+            if (formattedText.hasMultipleUniqueSections()) {
+                updatePaintProperties = [&, currentSectionIndex = optional<std::size_t>{}](std::size_t symbolSectionIndex, bool updateLastSection) mutable {
+                    if (currentSectionIndex && (updateLastSection || *currentSectionIndex != symbolSectionIndex)) {
+                        const auto& formattedSection = sectionOptionsToValue(formattedText.sectionAt(*currentSectionIndex));
+                        for (auto& pair : bucket->paintProperties) {
+                            pair.second.textBinders.populateVertexVectors(feature, bucket->text.vertices.vertexSize(), {}, {}, formattedSection);
+                        }
+                    }
+                    currentSectionIndex = symbolSectionIndex;
+                };
+            } else {
+                updatePaintProperties = [&](std::size_t, bool updateLastSection) {
+                    if (updateLastSection) {
+                        const auto& formattedSection = sectionOptionsToValue(formattedText.sectionAt(0));
+                        for (auto& pair : bucket->paintProperties) {
+                            pair.second.textBinders.populateVertexVectors(feature, bucket->text.vertices.vertexSize(), {}, {}, formattedSection);
+                        }
+                    }
+                };
             }
 
-            if (symbolInstance.writingModes & WritingModeType::Vertical) {
+            const Range<float> sizeData = bucket->textSizeBinder->getVertexSizeData(feature);
+            auto addSymbolGlyphQuads = [&](WritingModeType writingMode,
```

Comment:
could be also a member function?

Reply:
`[&]` make it unclear what exactly has been captured

Reply:
done